force-app/main/default/classes/TestRunProcessor.cls

```diff
@@ -103,11 +103,40 @@ public with sharing class TestRunProcessor implements Schedulable {
                     Method_Pass__c = apexTestResult.Outcome == 'Pass' ? true : false,
                     Name = apexTestResult.ApexClass.Name,
                     Stack_Trace__c = apexTestResult.stackTrace,
-                    Test_Run__c = testRun.Id
+                    Test_Run__c = testRun.Id 
                 )
             );
+            
         }
-
-        insert results;
+        
+		List<Test_Run_Method_Result__c> result = processPassFailDates(results);
+        insert result;
+    }
+    
+    private List<Test_Run_Method_Result__c> processPassFailDates(List<Test_Run_Method_Result__c> result) {
+        List<String> failedTests = new List<String>();
+        for (Test_Run_Method_Result__c r : result) {
+            if (!r.Method_Pass__c) {
+                failedTests.add(r.Method_Name__c);
+            }
+        }      
+        List<Test_Run_Method_Result__c> failedCousins = [SELECT Id, Method_Name__c, Method_Pass__c, CreatedDate FROM Test_Run_Method_Result__c 
+                                                   WHERE Method_Name__c IN :failedTests ORDER BY CreatedDate DESC];
+        
+        for (Test_Run_Method_Result__c r : result) {
+            if (!r.Method_Pass__c) {
+                r.First_Failure__c = r.CreatedDate;
+                for (Test_Run_Method_Result__c f : failedCousins) {
```

Comment:
Instead of the nested for each, would be much more efficient to create a `Map<String, List<Test_Run_Method_Result__c>` where the key is the `Method_Name__c` ahead of time.  The you can just loop through each `result` and get all the matching test methods by calling `pastResults.get(r.Method_Name__c)`.

You'll still end up with nested loops but the second loop will be on a MUCH smaller collection.

Reply:
Ok. I added a map instead. Came up with the same results so that's good. Had to add another for loop to create the map. Couldn't come up with a more efficient way, but maybe you have one in mind

Reply:
nope, this is the most efficient.  With this code you only see each item at most twice (once to create the map and when processing the failures).  Your previous implementation would visit each item on every failure (worst case).